test/e2e/federated-ingress.go

```diff
@@ -484,3 +513,119 @@ func getFederatedIngressAddress(client *fedclientset.Clientset, ns, name string)
 	}
 	return addresses, nil
 }
+
+// getNodeNames returns the names of nodes in a cluster.
+func getZoneNodesMap(clientset *kubeclientset.Clientset) (map[string][]string, error) {
+	zoneNodes := make(map[string][]string)
+	nodes, err := clientset.Core().Nodes().List(v1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+	// Only consider schedulable nodes, because we won't
+	// have backends in non-schedulable nodes and don't
+	// want to open port to those nodes.
+	// Also, master node is unschedulable and contains
+	// a different tag than all other nodes, so there
+	// is no point in considering that.
+	for _, n := range nodes.Items {
+		if n.Spec.Unschedulable {
+			continue
+		}
+		zone, ok := n.Labels[metav1.LabelZoneFailureDomain]
+		if !ok {
+			framework.Logf("skipping node %q from ZoneNode map because no zone label was found", n.Name)
+			continue
+		}
+		nodes, ok := zoneNodes[zone]
+		if !ok {
+			nodes = make([]string, 0)
+		}
+		nodes = append(nodes, n.Name)
+		zoneNodes[zone] = nodes
+	}
+	return zoneNodes, nil
+}
+
+func updateMap(merged, src map[string][]string) {
+	for zone, nodes := range src {
+		mNodes, ok := merged[zone]
+		if !ok {
+			mNodes = make([]string, 0)
+		}
+		mNodes = append(mNodes, nodes...)
+		merged[zone] = mNodes
+	}
+}
+
+// createFederatedIngressFirewallRule creates a firewall rule for the
+// federated ingress allowing traffic to the service shards in all the
+// underlying clusters.
+// TODO(madhusudancs): Remove this once
+// https://github.com/kubernetes/kubernetes/issues/37306
+// is implemented.
+func createFederatedIngressFirewallRules(ns, ingressName string, clusters map[string]*cluster, svc *v1.Service) (map[string]string, error) {
+	mergedZoneNodes := make(map[string][]string)
+	for clusterName, cluster := range clusters {
+		zoneNodes, err := getZoneNodesMap(cluster.Clientset)
+		if err != nil {
+			return nil, err
+		}
+		framework.Logf("nodes to zones map for cluster %q: %#v", clusterName, mergedZoneNodes)
```

Comment:
Why? We rarely use glog in e2e test code. It is incredibly useful while debugging in a non-interactive run, meaning jenkins.